lib/backend/etcdv3/etcdv3.go

```diff
@@ -0,0 +1,462 @@
+		if len(getResp.Kvs) == 0 {
+			logCxt.Info("Update transaction failed due to resource not existing")
+			return nil, errors.ErrorResourceDoesNotExist{Identifier: d.Key}
+		}
+
+		logCxt.Info("Update transaction failed due to resource update conflict")
+		var existing *model.KVPair
+		if v, err := model.ParseValue(d.Key, getResp.Kvs[0].Value); err == nil {
+			logCxt.Debug("Parsed current entry to return in response")
+			existing = &model.KVPair{
+				Key:      d.Key,
+				Value:    v,
+				Revision: strconv.FormatInt(getResp.Kvs[0].ModRevision, 10),
+			}
+		}
+		return existing, errors.ErrorResourceUpdateConflict{Identifier: d.Key}
+	}
+
+	d.Revision = strconv.FormatInt(txnResp.Header.Revision, 10)
+
+	return d, nil
+}
+
+// Apply an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdV3Client) Apply(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Apply request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+
+	logCxt.Debug("Performing etcdv3 Put for Apply request")
+	resp, err := c.etcdClient.Put(context.Background(), key, value)
+	if err != nil {
+		logCxt.WithError(err).Warning("Apply failed")
+		return nil, errors.ErrorDatastoreError{Err: err}
+	}
+
+	d.Revision = strconv.FormatInt(resp.Header.Revision, 10)
+
+	return d, nil
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
+func (c *EtcdV3Client) Delete(k model.Key, revision string) error {
+	logCxt := log.WithFields(log.Fields{"model-key": k, "rev": revision})
+	logCxt.Debug("Processing Delete request")
+	key, err := model.KeyToDefaultDeletePath(k)
+	if err != nil {
+		return err
+	}
+	logCxt = logCxt.WithField("etcdv3-key", key)
+
+	conds := []clientv3.Cmp{}
+	if len(revision) != 0 {
+		rev, err := strconv.ParseInt(revision, 10, 64)
+		if err != nil {
+			logCxt.Info("Unable to parse Revision")
+			return err
+		}
+		conds = append(conds, clientv3.Compare(clientv3.ModRevision(key), "=", rev))
+	}
+
+	// Perform the delete transaction - note that this is an exact delete, not a prefix delete.
+	logCxt.Debug("Performing etcdv3 transaction for Delete request")
+	txnResp, err := c.etcdClient.Txn(context.Background()).If(
+		conds...,
+	).Then(
+		clientv3.OpDelete(key),
+	).Commit()
+	if err != nil {
+		logCxt.WithError(err).Warning("Delete failed")
+		return errors.ErrorDatastoreError{Err: err, Identifier: k}
+	}
+
+	if !txnResp.Succeeded {
+		logCxt.Info("Delete transaction failed due to resource update conflict")
+		return errors.ErrorResourceUpdateConflict{Identifier: k}
+	}
+
+	delResp := txnResp.Responses[0].GetResponseDeleteRange()
+	if delResp.Deleted == 0 {
+		logCxt.Info("Delete transaction failed due resource not existing")
+		return errors.ErrorResourceDoesNotExist{Identifier: k}
```

Comment:
so does this mean, when the resource doesn't exist and you try to delete it, you still get `txnResp.Succeeded` = `true`??

Reply:
Yes I believe that is the case, and the UTs would suggest that it correct too.  Checking the number of deleted entries is what indicates whether the entry existed in the first place.

Again - it would make sense for you to dive into the etcdv3 calls too to make sure I am not misusing them.

Reply:
I'm going to add a commit to beef up the e2e UT/FVs to check the actual error contents.